source/common/network/peer_buffer.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * The interface for the writer.
+ */
+class WritablePeer {
+public:
+  virtual ~WritablePeer() = default;
+
+  /**
+   * Set the flag to indicate no further write from peer.
+   */
+  virtual void setWriteEnd() PURE;
+  virtual bool isWriteEndSet() PURE;
+
+  /**
+   * Raised when peer is destroyed. No further write to peer is allowed.
+   */
+  virtual void onPeerDestroy() PURE;
+
+  /**
+   * Notify that consumable data arrives. The consumable data can be either data to read, or the end
+   * of stream event.
+   */
+  virtual void maybeSetNewData() PURE;
+
+  /**
+   * @return the buffer to be written.
+   */
+  virtual Buffer::Instance* getWriteBuffer() PURE;
+
+  /**
+   * @return false more data is acceptable.
+   */
+  virtual bool isWritable() const PURE;
+
+  /**
+   * @return true if peer is valid and writable.
+   */
+  virtual bool isPeerWritable() const PURE;
+
+  /**
+   * Raised by the peer when the peer switch from high water mark to low.
+   */
+  virtual void onPeerBufferWritable() PURE;
+};
+
+/**
+ * The interface for the buffer owner who want to consume the buffer.
```

Comment:
What does this mean?

Reply:
The buffer has two group of use case, one as writer which probe the states of buffer and want to populate the data.

The others are reader who will query the buffer and drain the buffer.

Ideally the writer is another peer BufferedIoSocketHandle which refers to this buffer, and 
the reader is the BufferedIoSocketHandle which owns this buffer.

This interface is not used independently, I am going to remove this interface. WDTY?

Reply:
Yes, please remove this interface.